### ISSUE TITLE
qwen3_5.jinja: Assumes String Content on First System Message; Multimodal Assumes Lists — Breaks HF Datasets JSON/PyArrow for VL JSONL

### ISSUE THREAD

**Issue by NicoliAraujo** [axolotl-ai-cloud/axolotl#3590]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Summary

There is a mismatch between:

1. **Multimodal dataset documentation** — messages use OpenAI-style **`content` as a list** of parts (including for `system`), which is the natural shape for vision turns.
2. **`qwen3_5.jinja` (no-tools branch)** — the first `system` message is rendered with **string concatenation** on `messages[0].content`, which assumes a **string**, not a list.
3. **Hugging Face `datasets` + JSON** — PyArrow infers a **single type** for `messages[].content` across all messages in the file. If `system` is a string and `user` is a list, `load_dataset("json", ...)` fails with `ArrowInvalid` (e.g. *Column(/messages/[]/content) changed from string to array*).

Together, users following the multimodal doc for Qwen3.5 VL hit either template errors (list `system`) or dataset load errors (string `system` + list `user`).

### Documentation reference

Axolotl multimodal docs describe `content` as a list of parts and show **`system` with list `content`**:

https://docs.axolotl.ai/docs/multimodal.html  

For Qwen3.5 the docs recommend `chat_template: qwen3_5`.

### Related Issues

Hugging Face `datasets` / Arrow limitation on heterogeneous nested JSON:

https://github.com/huggingface/datasets/issues/7322  

### Current behaviour

In `src/axolotl/utils/chat_templates/templates/qwen3_5.jinja`:

```jinja
{%- if messages[0].role == 'system' %}
    {{- '<|im_start|>system\n' + messages[0].content + '<|redacted_im_end|>\n' }}
{%- endif %}
```

For `user` / non-first `system`, the same template correctly branches on `message['content'] is string` vs iterating list parts for vision tokens.

### Minimal data shape examples

**Fails `load_dataset("json")` (mixed `content` types in one row):**

```json
{
  "messages": [
    {"role": "system", "content": "You are helpful."},
    {"role": "user", "content": [{"type": "text", "text": "Hi"}, {"type": "image", "path": "/tmp/x.jpg"}]}
  ]
}
```

**Loads with Arrow if every `content` is a list, but `qwen3_5` first-system line concatenates list + string (template error):**

```json
{
  "messages": [
    {"role": "system", "content": [{"type": "text", "text": "You are helpful."}]},
    {"role": "user", "content": [{"type": "text", "text": "Hi"}, {"type": "image", "path": "/tmp/x.jpg"}]}
  ]
}
```

### Steps to Reproduce

**A — `datasets` / PyArrow (`ArrowInvalid` on mixed `content` types)**

1. Create `repro_mixed.jsonl` with a single line containing the JSON object from **Fails `load_dataset("json")`** above (first minimal example: `system` with string `content`, `user` with list `content`).
2. Run:

   ```python
   from datasets import load_dataset

   load_dataset("json", data_files="repro_mixed.jsonl", split="train")
   ```

3. **Observed:** `pyarrow.lib.ArrowInvalid` (e.g. `Column(/messages/[]/content) changed from string to array in row 0`, or similar wording depending on `datasets` / PyArrow versions).

**B — `qwen3_5.jinja` first `system` branch (list `content` on `messages[0]`)**

1. Create `repro_all_list.jsonl` with a single line containing the JSON object from **Loads with Arrow if every `content` is a list** above (both `system` and `user` use list-shaped `content`).
2. Confirm `load_dataset("json", data_files="repro_all_list.jsonl", split="train")` succeeds (or `json.loads` the line into a `messages` list).
3. Load a Qwen3.5 processor and the same `qwen3_5` chat template Axolotl ships (e.g. copy `src/axolotl/utils/chat_templates/templates/qwen3_5.jinja` into `tokenizer.chat_template`, or use Axolotl’s tokenizer wiring for `chat_template: qwen3_5`).
4. Call `apply_chat_template` on `messages` with `tokenize=False` (or the Axolotl code path that renders the template for a batch row).

5. **Observed:** template / Jinja failure from treating `messages[0].content` as a string in the no-tools first-system branch (e.g. attempt to concatenate a list with `'<|im_start|>system\n'`).

### Config yaml

```yaml
base_model: Qwen/Qwen3.5-2B
base_model_config: Qwen/Qwen3.5-2B
tokenizer_config: Qwen/Qwen3.5-2B
processor_type: AutoProcessor

sequence_len: 13056
pad_to_sequence_len: true
train_on_inputs: false
skip_prepare_dataset: true
remove_unused_columns: false
sample_packing: false

chat_template: qwen3_5
chat_template_kwargs:
  enable_thinking: false

datasets:
  - path: training_data.jsonl
    type: chat_template
    field_messages: messages
    roles_to_train: ["assistant"]
    split: train

val_set_size: 0
dataset_prepared_path: cache/training-data-qwen3.5-2b
output_dir: outputs/training-data-qwen3.5-2b

# === HYPERPARAMS (full training, same as v3-8gpu) ===
micro_batch_size: 1
gradient_accumulation_steps: 2
num_epochs: 4
warmup_ratio: 0.1
optimizer: adamw_torch_fused
learning_rate: 2e-5
eval_strategy: "no"
saves_per_epoch: 1
weight_decay: 0.0
save_only_model: false

unfrozen_parameters:
  - model.language_model.*

# === GPU / RUNTIME ===
load_in_8bit: false
load_in_4bit: false
bf16: true
tf32: true
deepspeed: deepspeed/zero2-gpu-only.json

wandb_project: project
wandb_name: run
logging_steps: 1

dataloader_num_workers: 0
dataloader_pin_memory: true
ddp_timeout: 7200

plugins:
  - axolotl.integrations.liger.LigerPlugin
liger_rope: false
liger_rms_norm: true
liger_glu_activation: true
liger_fused_linear_cross_entropy: true
flash_attention: true
resume_from_checkpoint:
gradient_checkpointing: true
gradient_checkpointing_kwargs:
  use_reentrant: false

# Run with 8 GPUs:
# export AXOLOTL_NCCL_TIMEOUT=7200
# torchrun --nproc_per_node=8 -m axolotl.cli.train qwen3.5-2b-qwen3_5.yaml
```

### Possible solution

One of the following would resolve the inconsistency:

- **A)** Extend the no-tools (and tools) branch so the **first** `system` message uses the same **string vs list** rendering as later `user`/`system` turns (only text parts for system; images remain invalid as today if desired), **or**
- **B)** Document explicitly that for `chat_template: qwen3_5`, the **first** `system` **`content` must be a string** and that **homogeneous JSON typing** for `datasets` requires either all-string or all-list `content` across messages — and align the multimodal examples accordingly.

Option **A** aligns the template with the published multimodal message format and allows **all-list `content`** in JSONL, which plays well with `load_dataset("json")`.

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

v0.15.0

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.
